### PR TITLE
Run3-hcx373 Correct the HcalTopology class for 2 new geometry modes

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Correct the HcalTopology class for 2 new geometry modes. Geometry mode is used to specify the detector scenario. LHC was the fixed version of the legacy detector. New modes are defined to specify the current situation with new ZDC assignment with the RPD detector (Run3) and the Phase2 setup (Run4) with no HE detector. It will be expanded in future to mark when HF DetIds were enlarged (2016), HE started expansion (2017) and HB had several depths (post LS2). Some of this is reflected in this version of HcalTopology. The usage of the modes was not correctly interpreted in the earlier version of the HcalTopology code.

#### PR validation:

Use the runTheMatrix test workflows

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Backport to 14_1_X using #46309

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by Dr15Jones** [comment]
It would seem to me that instead of having `if ( mode_ == .... || mode_ ....)` all over the code, to use a different new variable `phase_` which is either Phase1 or Phase2 based on the value of mode_.

**Comment by bsunanda** [comment]
Thanks Chris - trying to follow your suggestion

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 1]
```diff
@@ -940,8 +929,7 @@ bool HcalTopology::decrementDepth(HcalDetId& detId) const {
       }
     }
   } else if ((subdet == HcalEndcap) && (etaRing == lastHERing()) && (depth == hcons_->getDepthEta29(iphi, zside, 0)) &&
-             (mode_ != HcalTopologyMode::SLHC) && (mode_ != HcalTopologyMode::Run3) &&
-             (mode_ != HcalTopologyMode::Run4)) {
+             phase2()) {
```
Comment (Dr15Jones) [inline Geometry/CaloTopology/src/HcalTopology.cc:932]: `!phase2()` ?

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 2]
```diff
@@ -896,17 +888,14 @@ bool HcalTopology::incrementDepth(HcalDetId& detId) const {
...
-    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing() - 1) && (mode_ != HcalTopologyMode::SLHC) &&
-               (mode_ != HcalTopologyMode::Run3) && (mode_ != HcalTopologyMode::Run4)) {
+    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing() - 1) && phase2()) {
       // guard ring HF29 is behind HE 28
       subdet = HcalForward;
       (ieta > 0) ? ++ieta : --ieta;
       depth = 1;
-    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing()) && (mode_ != HcalTopologyMode::SLHC) &&
-               (mode_ != HcalTopologyMode::Run3) && (mode_ != HcalTopologyMode::Run4)) {
+    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing()) && phase2()) {
```
Comment (Dr15Jones) [inline Geometry/CaloTopology/src/HcalTopology.cc:898]: `!phase2()`?

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 3]
```diff
@@ -896,17 +888,14 @@ bool HcalTopology::incrementDepth(HcalDetId& detId) const {
     } else if (subdet == HcalBarrel && etaRing == lastHBRing()) {
       // overlap
       subdet = HcalEndcap;
-      if ((mode_ == HcalTopologyMode::SLHC) || (mode_ == HcalTopologyMode::H2HE) || (mode_ == HcalTopologyMode::Run3) ||
-          (mode_ == HcalTopologyMode::Run4))
+      if (phase2() || (mode_ == HcalTopologyMode::H2HE))
         depth = hcons_->getDepthEta16(2, iphi, zside);
-    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing() - 1) && (mode_ != HcalTopologyMode::SLHC) &&
-               (mode_ != HcalTopologyMode::Run3) && (mode_ != HcalTopologyMode::Run4)) {
+    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing() - 1) && phase2()) {
```
Comment (Dr15Jones) [inline Geometry/CaloTopology/src/HcalTopology.cc:893]: `!phase2()`?

Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc]: Thanks - done

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 2 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc]: Thanks - taken care of now

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc]: Thanks taken care of

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by makortel** [comment]
The "small change" commit seems to change the values for `Run3`. How do we know which values (before or after the commit) are correct?

**Comment by bsunanda** [comment]
This is the final version that is used in the creation of the payloads of 2024 geometry. So in some way this is a validated version.

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 4]
```diff
@@ -195,18 +195,14 @@ HcalTopology::HcalTopology(HcalTopologyMode::Mode mode,
       HFSize_(kHFSizePreLS1),
       HTSize_(kHTSizePreLS1),
       CALIBSize_(kCALIBSizePreLS1),
-      numberOfShapes_(
-          ((mode == HcalTopologyMode::SLHC) || (mode_ == HcalTopologyMode::Run3) || (mode_ == HcalTopologyMode::Run4))
-              ? 500
-              : 87) {
+      numberOfShapes_(phase2() ? 500 : 87) {
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:198]: This does not correspond to [L82](https://github.com/cms-sw/cmssw/blob/eb00e8ccf01deca0dec95e877569ecf300535cfb/Geometry/CaloTopology/src/HcalTopology.cc#L82). Just checking: is it correct?

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 5]
```diff
@@ -571,8 +567,7 @@ bool HcalTopology::validRaw(const HcalDetId& id) const {
 
   if (ok) {
     if (subdet == HcalBarrel) {
-      if ((mode_ == HcalTopologyMode::SLHC) || (mode_ == HcalTopologyMode::H2HE) || (mode_ == HcalTopologyMode::Run3) ||
-          (mode_ == HcalTopologyMode::Run4)) {
+      if (phase2() || (mode_ == HcalTopologyMode::H2HE)) {
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:570]: This does not include Run3 now: is it correct? (Just checking)

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 6]
```diff
@@ -581,8 +576,7 @@ bool HcalTopology::validRaw(const HcalDetId& id) const {
           ok = false;
       }
     } else if (subdet == HcalEndcap) {
-      if ((mode_ == HcalTopologyMode::SLHC) || (mode_ == HcalTopologyMode::H2HE) || (mode_ == HcalTopologyMode::Run3) ||
-          (mode_ == HcalTopologyMode::Run4)) {
+      if (phase2() || (mode_ == HcalTopologyMode::H2HE)) {
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:579]: This does not include Run3 now: is it correct? (Just checking)

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 7]
```diff
@@ -823,8 +817,7 @@ int HcalTopology::decAIEta(const HcalDetId& id, HcalDetId neighbors[2]) const {
 void HcalTopology::depthBinInformation(
     HcalSubdetector subdet, int etaRing, int iphi, int zside, int& nDepthBins, int& startingBin) const {
   if (subdet == HcalBarrel) {
-    if ((mode_ == HcalTopologyMode::SLHC) || (mode_ == HcalTopologyMode::H2HE) || (mode_ == HcalTopologyMode::Run3) ||
-        (mode_ == HcalTopologyMode::Run4)) {
+    if (phase2() || (mode_ == HcalTopologyMode::H2HE)) {
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:820]: This does not include Run3 now: is it correct? (Just checking)

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 8]
```diff
@@ -896,17 +888,14 @@ bool HcalTopology::incrementDepth(HcalDetId& detId) const {
     } else if (subdet == HcalBarrel && etaRing == lastHBRing()) {
       // overlap
       subdet = HcalEndcap;
-      if ((mode_ == HcalTopologyMode::SLHC) || (mode_ == HcalTopologyMode::H2HE) || (mode_ == HcalTopologyMode::Run3) ||
-          (mode_ == HcalTopologyMode::Run4))
+      if (phase2() || (mode_ == HcalTopologyMode::H2HE))
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:891]: This does not include Run3 now: is it correct? (Just checking)

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 9]
```diff
@@ -896,17 +888,14 @@ bool HcalTopology::incrementDepth(HcalDetId& detId) const {
...
-    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing() - 1) && (mode_ != HcalTopologyMode::SLHC) &&
-               (mode_ != HcalTopologyMode::Run3) && (mode_ != HcalTopologyMode::Run4)) {
+    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing() - 1) && !phase2()) {
       // guard ring HF29 is behind HE 28
       subdet = HcalForward;
       (ieta > 0) ? ++ieta : --ieta;
       depth = 1;
-    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing()) && (mode_ != HcalTopologyMode::SLHC) &&
-               (mode_ != HcalTopologyMode::Run3) && (mode_ != HcalTopologyMode::Run4)) {
+    } else if ((subdet == HcalEndcap) && (etaRing == lastHERing()) && !phase2()) {
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:898]: Please check

File: Geometry/CaloTopology/src/HcalTopology.cc [hunk 10]
```diff
@@ -940,8 +929,7 @@ bool HcalTopology::decrementDepth(HcalDetId& detId) const {
       }
     }
   } else if ((subdet == HcalEndcap) && (etaRing == lastHERing()) && (depth == hcons_->getDepthEta29(iphi, zside, 0)) &&
-             (mode_ != HcalTopologyMode::SLHC) && (mode_ != HcalTopologyMode::Run3) &&
-             (mode_ != HcalTopologyMode::Run4)) {
+             !phase2()) {
```
Comment (perrotta) [inline Geometry/CaloTopology/src/HcalTopology.cc:932]: Again, `HcalTopologyMode::Run3` in not in your definition of `phase2()` as in [L175](https://github.com/cms-sw/cmssw/pull/46308/files#diff-bac5fd17f1d22b6df2fcd5469c58d31282ed972fc44d7f4c5c849c111c8b8441R175)

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 6 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc:579]: Yes. The logic is with or without HE

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 5 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc:570]: Yes. Earlier logic was wrong

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 4 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc]: It was more or less correct. Updated it with mode LHC or not

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 7 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc:820]: Yes it is

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 9 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc:898]: Done

File: Geometry/CaloTopology/src/HcalTopology.cc (hunk 10 above)
Comment (bsunanda) [reply Geometry/CaloTopology/src/HcalTopology.cc:932]: Run3 is not a part of Phase2

**Comment by bsunanda** [comment]
+geometry

**Review by perrotta — COMMENTED** [review]
This PR includes some updates/fixes with respect to the HcalTopology currently in CMSSW: please add the description of the rationale of those changes/fixes in the PR description, so that it can also remain as a future documentation (besides helping now the poor reviewer of the PR).

As far as I understand, the fixes are the following ones:
- (1) Avoid undefined behaviour at [L75]( https://github.com/cms-sw/cmssw/blob/d0937b801adccf41baf579df4b2230cba7e23442/Geometry/CaloTopology/src/HcalTopology.cc#L75) for modes different from `LHC` and `SLHC`
- (2) Remove `Run3` from the Phase2 modes, as it is erroneously now the case in the release

Then I remain with my doubt about [L198]( https://github.com/cms-sw/cmssw/blob/d0937b801adccf41baf579df4b2230cba7e23442/Geometry/CaloTopology/src/HcalTopology.cc#L198) that I expect you can explain here (or fix, if needed)

**Comment by Dr15Jones** [comment]
So I removed the constructor
[```](https://github.com/cms-sw/cmssw/blob/9668a4de07a525a538f0dd532c3e263d1a2d3e78/Geometry/CaloTopology/interface/HcalTopology.h#L29-L32)

from the class, did `git cms-checkdeps -a` and rebuilt all the relevant parts of CMSSW and everything compiled just fine. That means we should remove that constructor for the class to help with maintenance.

**Comment by perrotta** [comment]
> So I removed the constructor
> 
> https://github.com/cms-sw/cmssw/blob/9668a4de07a525a538f0dd532c3e263d1a2d3e78/Geometry/CaloTopology/interface/HcalTopology.h#L29-L32
> 
> from the class, did `git cms-checkdeps -a` and rebuilt all the relevant parts of CMSSW and everything compiled just fine. That means we should remove that constructor for the class to help with maintenance.
> 
> The other reason to remove that constructor is it can lead to an instance of HcalTopology which will crash on certain calls because large number of member data expect the member data hcons_ to be set, which does not happen with that constructor.

Thank you @Dr15Jones , I second your proposal.
It would also remove the ambiguity I pointed out in https://github.com/cms-sw/cmssw/pull/46308#discussion_r1795094716 : still I would like to hear from @bsunanda which is the correct formulation for `numberOfShapes_`

**Comment by bsunanda** [comment]
@perrotta Please sign this. I am preparing the correct implementation of HcalTopology given what I wrote earlier for different eras of HCAL. This version is adequate for CMSSW_14_1_X.

**Comment by perrotta** [comment]
> @perrotta Please sign this. I am preparing the correct implementation of HcalTopology given what I wrote earlier for different eras of HCAL. This version is adequate for CMSSW_14_1_X.

@bsunanda could you please at least reply to my question in https://github.com/cms-sw/cmssw/pull/46308#discussion_r1795094716 ? Since there are two different computations of the same quantity in the two different constructors I think it is mandatory to understand whether it is correct that they are such different, and if not the wrong one must be fixed (or even removed, as suggested by @Dr15Jones )

**Comment by bsunanda** [comment]
@perrotta At the moment the second constructor is not used and it won't create a mess. I shall surely remove it in the next version. It was left there because when we were changing the HcalTopology (some 10-12 years ago), some of the codes were still using HcalTopology without accessing Hcal..Const from EventSetup. That practice has disappeared as seen by Chris's test. So it is safe to remove it. If I remove this in this PR, this will trigger the tests and the whole process of integration 
will be delayed.

**Comment by perrotta** [comment]
> @perrotta At the moment the second constructor is not used and it won't create a mess. I shall surely remove it in the next version. It was left there because when we were changing the HcalTopology (some 10-12 years ago), some of the codes were still using HcalTopology without accessing Hcal..Const from EventSetup. That practice has disappeared as seen by Chris's test. So it is safe to remove it. If I remove this in this PR, this will trigger the tests and the whole process of integration will be delayed.

Fine.
But which is the correct definition of `numberOfShapes_`  (fourth time I ask here....)? The one in the constructor that is used, or the one in the unused constructor?
Only if the correct definition is the one in the used constructor I can accept:
- signing and merging this PR in the master
- make at once another PR in master that removes the unused (and bugged) constructor
- backport in #46309 the combination of the two PRs

**Comment by bsunanda** [comment]
@perrotta The first constructor is the right one. I shall remove the second constructor in my next PR

**Comment by perrotta** [comment]
+db
- This PR implements some necessary fix to the HcalTopology class:
  - Avoid undefined behaviour at [L75](https://github.com/cms-sw/cmssw/blob/d0937b801adccf41baf579df4b2230cba7e23442/Geometry/CaloTopology/src/HcalTopology.cc#L75) for modes different from LHC and SLHC
  - Remove `Run3` from the Phase2 modes, as it is erroneously now the case in the release
-  It is understood that another (minor) fix is needed to remove a bugged constructor: a fix will be also provided and combined in the backport PR

**Comment by perrotta** [comment]
urgent
(@cms-sw/orp-l2 could you please merge in time for the 1100 IB, so that the necessary actions can be taken at once to implement further fixes and prepare a backport?)

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**